### DISCUSSION TITLE
init fails with latest update

### DISCUSSION THREAD

**Discussion by dasoju** [magit/transient#295] [Support] [answered]
hi
emacs init fails with the following error
Symbol's function definition is void: static-if

i see its erroring out due to following in the transient package
transient.el:(static-if (>= emacs-major-version 30) ;transient--wrap-command

Not sure where static-if is defined, I am using emacs 29.1 (mitsuharu version)

**Comment by tarsius** [comment, marked as answer]
Update: This might be enough to fix the issue.

1. Uninstall `compat`, `transient` and all packages that depend on `transient`.
2. Exit Emacs.
3. Restart Emacs.
4. Reinstall the packages.

----

`static-if` is new in Emacs 30. For released versions of Emacs it is available in the `compat` package, version 30.0.0.0. `transient` depends on that version of `compat`.

When you updated `transient`, your package manager should also have updated `compat`.

Maybe it did, and all you have to do is to restart Emacs.

Maybe it didn't, and you have to manually update.

(What package manager are you using?)

**Reply by dasoju** [reply]
Reinstalling transient resolved the issue.

My original installation was installing transient as a dependency and during package autoupdate compact was deleted.

I have following set in my init
(setq auto-package-update-delete-old-versions t)
Not sure if this is triggering deletion of compat, which i was not expecting.

Thanks for the help

**Reply by a-daniel-eliason** [reply]
@tarsius , I am running Emacs 30.0.50, and I have 'compat' 30.0.0.0 and I have the same problem that 'static-if' does not exist.  To get 'magit' back, I had to edit 'static-if' back into my local copy of the source for 'transient'.  Other packages were broken, too.

I see that the source of 'compat' does indeed have some decl for 'static-if', but somehow it is not compiling through the normal 'list-packages' type of installation.

**Reply by a-daniel-eliason** [reply]
@tarsius, fyi, with Emacs 30.0.50 when I use 'list-packages' to install first 'compat' and then 'transient', this is what happens.  It looks like the right files of 'compat' are compiled (-30.el" in particular, which has 'static-if'), but 'static-if' is still not available to 'transient':
`
Compiling internal form(s) at Thu Jul 18 18:12:23 2024
Leaving directory ‘/home/eliason/.emacs.d/elpa/compat-30.0.0.0’

Compiling file /home/eliason/.emacs.d/elpa/compat-30.0.0.0/compat-25.el at Thu Jul 18 18:12:23 2024
Entering directory ‘/home/eliason/.emacs.d/elpa/compat-30.0.0.0/’

Compiling file /home/eliason/.emacs.d/elpa/compat-30.0.0.0/compat-26.el at Thu Jul 18 18:12:23 2024

Compiling file /home/eliason/.emacs.d/elpa/compat-30.0.0.0/compat-27.el at Thu Jul 18 18:12:23 2024

Compiling file /home/eliason/.emacs.d/elpa/compat-30.0.0.0/compat-28.el at Thu Jul 18 18:12:23 2024

Compiling file /home/eliason/.emacs.d/elpa/compat-30.0.0.0/compat-29.el at Thu Jul 18 18:12:23 2024

Compiling file /home/eliason/.emacs.d/elpa/compat-30.0.0.0/compat-30.el at Thu Jul 18 18:12:23 2024

Compiling file /home/eliason/.emacs.d/elpa/compat-30.0.0.0/compat.el at Thu Jul 18 18:12:23 2024

Compiling internal form(s) at Thu Jul 18 18:12:54 2024
Leaving directory ‘/home/eliason/.emacs.d/elpa/transient-0.7.3’

Compiling file /home/eliason/.emacs.d/elpa/transient-0.7.3/transient.el at Thu Jul 18 18:12:54 2024
Entering directory ‘/home/eliason/.emacs.d/elpa/transient-0.7.3/’

In end of data:
transient.el:2438:2: Warning: the function ‘static-if’ is not known to be
    defined.
`

**Reply by tarsius** [reply]
> I am running Emacs 30.0.50, and I have 'compat' 30.0.0.0 and I have the same problem that 'static-if' does not exist.

Based on that version string, we know that you are using an outdated development version of Emacs. The current version string, if you are using the `emacs-30` branch, is 30.0.60. If you are using the `master` branch, it is 31.0.50.

In all likelihood you are using a temporary snapshot from before `static-if` was added to Emacs. The `compat` package performs a version check to determine whether it does or does not need to define `static-if` (and other things new in what will become Emacs 30.1). But this version check is not precise enough to make that determination down to the commit level. So you will have to update your Emacs build.

The more general takeaway from this is, that if you use a development version of Emacs and run into issues, you must always first update Emacs.

**Reply by a-daniel-eliason** [reply]
Thanks for discussion.  I think all these troubles are due to the 'compat' package not being very sensitive to the version numbers.  My own solution for now is re-inserting 'static-if' to my local copy of 'transient', and my eventual solution is to recompile more recent Emacs, where I see that 'static-if' is included in "subr.el".

**Reply by bjacquet** [reply]
Ran into this issue last week.  I'm using Emacs 29.4 and managed to fix it by removing and re-installing compat, transient, and forge.

Thanks for your help suggesting this. :-)
